### DISCUSSION TITLE
关于Chrome 网上应用商店被拒绝相关事宜

### DISCUSSION THREAD

**Discussion by ronggang** [ronggang/PT-Plugin-Plus#318] [General]
近期，发布新版到 `Chrome 网上应用商店` 时，一直被 Goolge 拒绝，也没有明确给出拒绝的理由，经过多次邮件咨询终无果，也在网上看了一些相关内容：
- https://www.eviltester.com/blog/eviltester/2019/04/03-chrome-extension-rejection/
  - 这位老兄还录了视频

- uBlockOrigin 相关事件
  - https://blog.gslin.org/archives/2019/10/14/9244/ublock-origin-%E7%9A%84%E9%96%8B%E7%99%BC%E7%89%88-dev-%E8%A2%AB-chrome-web-store-%E6%8B%92%E7%B5%95%E7%9A%84%E4%BA%8B%E4%BB%B6/
  - https://github.com/uBlockOrigin/uBlock-issues/issues/745
  - https://www.reddit.com/r/chrome/comments/dgoymg/warning_ubo_ublock_origin_will_possibly_be/
  - https://hub.packtpub.com/after-backlash-for-rejecting-a-ublock-origin-update-from-the-chrome-web-store-google-accepts-ad-blocking-extension/

为了这个事情，前前后后大概有10多封邮件：
![image](https://user-images.githubusercontent.com/8065899/68728549-92c05180-0602-11ea-93c0-163639717a22.png)

Google发的内容都一样，像这样：
![image](https://user-images.githubusercontent.com/8065899/68728226-a7e8b080-0601-11ea-8fc3-00047aabf0c1.png)

我也尝试了做一些回复，但后来才发现，我在和机器人对话 🙄🙄🙄
![image](https://user-images.githubusercontent.com/8065899/68728766-2abe3b00-0603-11ea-9f3a-cd0d513d10f4.png)

起初以为没有隐私声明，但加了也一样 ，也尝试发布新版本，结果还是一样被拒绝：

![image](https://user-images.githubusercontent.com/8065899/68729010-c9e33280-0603-11ea-98f0-bfdf7b224d1e.png)

最后，我尝试通过申请恢复开发者账号的渠道反馈（ https://support.google.com/chrome_webstore/contact/cws_dev_appeals ），结果还是一个鸟样，估计还是机器回复
![image](https://user-images.githubusercontent.com/8065899/68729344-af5d8900-0604-11ea-8f88-29379aac9f16.png)

大概是意思是说已恢复，但结果还是被拒绝。

好吧，经过半个多月的折腾，心累，不想再折腾，如果Goolge还那鸟样，以后大家就用 zip 版吧，如果因为 Chrome 有烦人的提示，可以尝试使用其他没有提示的浏览器。

ronggang
2019.11.13

**Comment by luckywaa** [comment]
chrome怎么安装被拒绝的版本

**Comment by ted423** [comment]
@luckywaa 论wiki存在的意义

**Comment by GoldJohnKing** [comment]
你好，请问是否可以在 Microsoft 的 Edge Chromium 插件商店发布？微软的插件商店国内可免翻墙访问。

**Comment by ronggang** [comment]
> 你好，请问是否可以在 Microsoft 的 Edge Chromium 插件商店发布？微软的插件商店国内可免翻墙访问。

Edge Chromium 新版的市场还未完善，提交的插件半个月了还处于审核中。

**Comment by ronggang** [comment]
鉴于这段时间还是有同学会提起市场版什么时间发布，什么时候审核通过之类的，在此我把被拒绝的记录做以下整理：

## 2019-10-26 发布 [v1.4.0](https://github.com/ronggang/PT-Plugin-Plus/releases/tag/v1.4.0)
- 2019-10-30 被拒绝，原因同上：
![image](https://user-images.githubusercontent.com/8065899/70958617-c9333580-20b4-11ea-8b19-09486ff43f49.png)

## 2019-11-01 发布 [v1.4.2](https://github.com/ronggang/PT-Plugin-Plus/releases/tag/v1.4.2)
- 2019-11-03 被拒绝，原因同上
![image](https://user-images.githubusercontent.com/8065899/70958758-2e872680-20b5-11ea-8986-c9448698a5dc.png)
- 2019-11-11 重新发布一次，再次拒绝
![image](https://user-images.githubusercontent.com/8065899/70965740-bd06a280-20cb-11ea-82a5-514540640a02.png)

## 2019-11-22 发布 [v1.4.3](https://github.com/ronggang/PT-Plugin-Plus/releases/tag/v1.4.3)
- 本次发布前，尝试去除 `cookies` 功能后再进行发布
- 2019-11-26 再次被拒绝
![image](https://user-images.githubusercontent.com/8065899/70965865-1a9aef00-20cc-11ea-9be2-11a28eee228f.png)

## 2019-11-27 发布 [v1.4.4](https://github.com/ronggang/PT-Plugin-Plus/releases/tag/v1.4.4)
- 2019-11-29 再次被拒绝
![image](https://user-images.githubusercontent.com/8065899/70966917-ad895880-20cf-11ea-8e4c-0a47297454cc.png)
- 当天进行调整后重新发布
- 2019-12-03 再次被拒绝
![image](https://user-images.githubusercontent.com/8065899/70967686-ea564f00-20d1-11ea-9736-45b8ccad2882.png)

## 2019-12-09 发布 [v1.4.5](https://github.com/ronggang/PT-Plugin-Plus/releases/tag/v1.4.5)
- 2019-12-13 被拒绝
![image](https://user-images.githubusercontent.com/8065899/70965977-83826700-20cc-11ea-8889-2b0e2c716406.png)
- 2019-12-13 去除代码压缩和混淆再次上传至 CWS；
- 2019-12-17 再次被拒绝
![image](https://user-images.githubusercontent.com/8065899/70966737-3358d400-20cf-11ea-9f10-12b81d79c4d5.png)
![image](https://user-images.githubusercontent.com/8065899/70969358-68b4f000-20d6-11ea-98ee-f6678464d42a.png)

-----------------------------------------
至此，期间已先后邮件来往尽然达到 `40` 多封
![image](https://user-images.githubusercontent.com/8065899/70967164-79fafe00-20d0-11ea-9e89-6e5884454ca0.png)
细心的同学可能已注意到了，所谓的人工审核根本不存在，邮件回复时间基本在两个时间：
- 上午 12 点
- 下午 4 点

说白了还是机器在进行“审核”。

----------------------------------------
并非在下不努力，只是 `Google` 太大，我们太小，玩不起；感谢一起想办法的各位朋友；
如无特殊情况，本人将于 `2020-01-01` 从 `Chrome Web Store` 市场下架助手，不再向它提交新版本，再次感谢各位的支持，也请互相转告还在使用市场版但又有升级需求的同学，合理安排升级。

`2020-01-01` 之后如需更新您可以选择：
- [从 Microsoft Edge Addons 安装与更新](https://github.com/ronggang/PT-Plugin-Plus/install-from-microsoft-edge-addons)
- [从 Firefox Add-ons 安装与更新](https://github.com/ronggang/PT-Plugin-Plus/install-from-amo)
- [从 zip 安装与更新](https://github.com/ronggang/PT-Plugin-Plus/wiki/install-from-zip)
- [从 crx 安装与更新](https://github.com/ronggang/PT-Plugin-Plus/wiki/install-from-crx)


顺便 @dotproto ，简介是 Developer Advocate for Chrome Extensions.

---------------------------------

ronggang
2019-12-17

**Comment by ningning060705** [comment]
大神再想想办法吧，下架商店对于用户的影响真的挺大的，毕竟大部分用户都是在chrome上使用的吧

**Comment by ted423** [comment]
@ningning060705 可以的话你可以帮忙想出个办法来...

**Comment by ronggang** [comment]
> 大神再想想办法吧，下架商店对于用户的影响真的挺大的，毕竟大部分用户都是在chrome上使用的吧

并非在下不想办法，只因能力和精力有限，如有特殊通道还望告知。

**Comment by asoulist** [comment]
@ronggang 能否在国内其他chrome内核浏览器自建的扩展商店上架呢？比如大数字的

**Reply by potus2030** [reply]
可以的

**Comment by ted423** [comment]
@asoulist 我反对

**Comment by tongyifan** [comment]
@asoulist 反对+1

**Comment by ronggang** [comment]
> @ronggang 能否在国内其他chrome内核浏览器自建的扩展商店上架呢？比如大数字的

是指360吗？鉴于它的黑历史及没下限的作风，不会考虑和它产生任何关系。

**Reply by xkmttc** [reply]
360垃圾无底限，但凡有些善良的人都不愿意与它合作；

**Reply by potus2030** [reply]
其实谷歌也是50步笑一百步而已。为什么可以被谷歌虐？

**Comment by haig233** [comment]
@asoulist 反对

**Comment by dotproto** [comment]
Hey folks. I was just checking my github notifications and saw that I was tagged in this issue.

Apologies for the poor review experience. Normally in situations like this I try to explain the reason for the rejection, but unfortunately I'm not sure what went wrong with your extension. I'm going to reach out to my colleagues on the review team and ask them to take another look.

**Reply by maicss** [reply]
Any news?

**Reply by emooooooji** [reply]
hi. any news?

**Comment by ted423** [comment]
Wow

**Comment by xiuluo** [comment]
@dotproto 希望能有用。

**Comment by ronggang** [comment]
> Hey folks. I was just checking my github notifications and saw that I was tagged in this issue.
> 
> Apologies for the poor review experience. Normally in situations like this I try to explain the reason for the rejection, but unfortunately I'm not sure what went wrong with your extension. I'm going to reach out to my colleagues on the review team and ask them to take another look.

thanks for your help.

**Comment by ronggang** [comment]
@dotproto So "thank you" for your help, now I don't need to remove it by myself. 🙄

![image](https://user-images.githubusercontent.com/8065899/71427141-cefdcc00-26ef-11ea-8aee-ad64b9b23cbc.png)
![image](https://user-images.githubusercontent.com/8065899/71427166-1a17df00-26f0-11ea-8630-e25ac5a9fe18.png)

**Comment by ted423** [comment]
我在想问题是不是在于第三条..?

**Comment by ronggang** [comment]
> 我在想问题是不是在于第三条..?

猜不到，如果是那就不上了呗。

**Comment by Mosney** [comment]
不知道有没有计划开发其它形式的版本，比如做一个docker镜像，用户可以self-host，交互的话直接用WebUI可以利用现有的代码什么的

**Comment by jemyzhang** [comment]
@dotproto We urge to know the reason why the plugin is denied, this plugin is really very useful to us,  please help us, thanks again!

**Comment by ronggang** [comment]
> 不知道有没有计划开发其它形式的版本，比如做一个docker镜像，用户可以self-host，交互的话直接用WebUI可以利用现有的代码什么的

已在酝酿中，准备启用一个服务器版本，会和插件进行功能互补，搜索及一些需要后台支撑的功能会部署到服务器功能中。

**Comment by ted423** [comment]
@Mosney mos的想法果然与众不同

**Comment by jemyzhang** [comment]
> > 不知道有没有计划开发其它形式的版本，比如做一个docker镜像，用户可以self-host，交互的话直接用WebUI可以利用现有的代码什么的
> 
> 已在酝酿中，准备启用一个服务器版本，会和插件进行功能互补，搜索及一些需要后台支撑的功能会部署到服务器功能中。

要不要考虑telegram bot版本，直接搜索然后发送到本地服务器下载，爽的很，嘿嘿

**Comment by Rhilip** [comment]
> > 不知道有没有计划开发其它形式的版本，比如做一个docker镜像，用户可以self-host，交互的话直接用WebUI可以利用现有的代码什么的
> 
> 已在酝酿中，准备启用一个服务器版本，会和插件进行功能互补，搜索及一些需要后台支撑的功能会部署到服务器功能中。

个人对开发server持谨慎的态度。
1. 当初开发浏览器插件的原因之一是因为插件背景页调用XMLHTTPrequests方法可以直接携带浏览器已经有的cookies信息，从而不需要用户在另一软件中重新登录或复制粘贴cookies信息。
2. 就cs模式的，GitHub也已经有较为成熟的repo https://github.com/Jackett/Jackett ， 重写一边轮子似乎意义不大（Jackett也支持很多国内站点的搜索功能）

**Comment by Rhilip** [comment]
@Mosney 会考虑给我们发全薪的工资吗🤣

**Comment by ronggang** [comment]
> > > 不知道有没有计划开发其它形式的版本，比如做一个docker镜像，用户可以self-host，交互的话直接用WebUI可以利用现有的代码什么的
> > 
> > 
> > 已在酝酿中，准备启用一个服务器版本，会和插件进行功能互补，搜索及一些需要后台支撑的功能会部署到服务器功能中。
> 
> 个人对开发server持谨慎的态度。
> 
> 1. 当初开发浏览器插件的原因之一是因为插件背景页调用XMLHTTPrequests方法可以直接携带浏览器已经有的cookies信息，从而不需要用户在另一软件中重新登录或复制粘贴cookies信息。
> 2. 就cs模式的，GitHub也已经有较为成熟的repo https://github.com/Jackett/Jackett ， 重写一边轮子似乎意义不大（Jackett也支持很多国内站点的搜索功能）

是否真的实施还有待考量，最初的想法是这样的：
- 所有配置操作还在助手进行；
- 后端相当于部署一个“代理”，做以下事情：
  - 支撑助手的数据保存，提供 `api` 的方式更新数据，这样后台可启用数据库；
  - 做一些后台任务，比如定期更新数据等；站点 `cookies` 信息来源于已备份的数据；
  - 个人比较喜欢 `Kodi` ，希望和 `kodi` 数据库打通，这样可定期直接更新电影库及海报墙；
- 用户可按自己的需求自已部署后端；
- 助手依然是浏览器扩展的方式呈现，没有后端一样运行；

不过现在也只是想法阶段，有待商讨，具体事宜可另发主题进行讨论。

**Comment by ronggang** [comment]
> @Mosney 会考虑给我们发全薪的工资吗🤣

这可以有 👻

**Comment by Mosney** [comment]
> @Mosney 会考虑给我们发全薪的工资吗

我负责吹爆！走过路过的吃瓜群众请给R酱点个“吹爆”表情，谢谢！

**Comment by LsnmxNB** [comment]
[dmmjlmbkigbgpnjfiimhlnbnmppjhpea.zip](https://github.com/ronggang/PT-Plugin-Plus/files/3999804/dmmjlmbkigbgpnjfiimhlnbnmppjhpea.zip)

dmmjlmbkigbgpnjfiimhlnbnmppjhpea.zip 直接从Chrome复制出来的，扩展程序>>开发者模式>>加载已解压扩展程序就能用了


![Snipaste_2019-12-25_14-24-28](https://user-images.githubusercontent.com/42826027/71435069-4cdacb00-2722-11ea-8ba6-ff6866866f61.png)

**Comment by ted423** [comment]
@LsnmxNB 这个是可以用，甚至不用解压，但是文件夹被删了就不行了。

**Comment by LsnmxNB** [comment]
@ted423 PT-Plugin-Plus-v1.4.5.crx 直播丢扩展程序安装上无法打开，哈哈，对https://github.com/ronggang/PT-Plugin-Plus/wiki/supported-sites 已支持站点怎么都没有CMCT,擦导入了，CMCT改名了

**Comment by ted423** [comment]
@LsnmxNB https://github.com/ronggang/PT-Plugin-Plus/wiki/install-from-crx
CMCT就是SSD啊

**Comment by ted423** [comment]
@ted423 说反……ssd就是cmct

**Comment by LsnmxNB** [comment]
> @ted423 说反……ssd就是cmct

PT-Plugin-Plus看完官方文档没发现有自动查找新种刷流量的功能啊

**Comment by Mosney** [comment]
@LsnmxNB  @ted423  offtopic请另开主题讨论

**Comment by dotproto** [comment]
I'm on vacation this week but still trying to help. I just checked in on the request I submitted. Unfortunately I don't have an answer for you yet. I will continue to check on it periodically.

**Comment by ronggang** [comment]
> I'm on vacation this week but still trying to help. I just checked in on the request I submitted. Unfortunately I don't have an answer for you yet. I will continue to check on it periodically.

Thanks again, We look forward to good news.

**Comment by eiloogs** [comment]
烦人的弹窗使用.dev 或更高的开发版 就阔以了- -

**Comment by HansYeoh** [comment]
https://blog.csdn.net/qq_37933685/article/details/80504103

**Comment by HansYeoh** [comment]
https://blog.csdn.net/qq_37933685/article/details/80504103可以试试这个操作

**Comment by ywyha** [comment]
还是没有消息吗，有些郁闷，大佬们辛苦开发的优良插件无法上架，真的是太可惜了。

**Comment by orangexw** [comment]
今天重装系统刚好安装了EDGE，在微软扩展里查询PT-Plugin-Plus 已上架，大家多了一种选择。

**Comment by xkww3n** [comment]
挖个远古巨坟。不上架可能是在某个页面里引用了一个外站js？我在安装时提示这违反了什么chrome的规范。

**Reply by AaronConlon** [reply]
关键是有时候拒绝的理由太简单，一点蛛丝马迹都没有，想排查都无从下手

**Comment by ccxuy** [comment]
可以试试在userscript上上架..., 通过tempermonkey装到chrome里

**Comment by Rhilip** [comment]
> 可以试试在userscript上上架..., 通过tempermonkey装到chrome里

关于目前插件的 下载推送、聚合搜索、个人信息整理、辅种任务 几个模块，请问您有写userscript的思路或者简易实现吗？不然请参 #36

**Comment by alexcdever** [comment]
给一个macOS下的安装方法：
1. 下载作者发布的crx文件
2. 将该文件的后缀名从`.crx`改为`.tar.gz`
3. 将这个文件解压到一个指定的文件夹里
4. 在Chrome的扩展程序页面的右上角那里开启开发者模式
![image](https://user-images.githubusercontent.com/14200813/92153158-6528eb00-ee56-11ea-87a9-a9051274e29a.png)
5. 同样是在这个页面里的左上角有个`加载已解压的拓展程序`按钮，点击它并选中刚刚放置解压的文件的文件夹
6. 会弹框问是否安装扩展程序，点`是`，安装成功

**Comment by ted423** [comment]
@alexcdever Mac版最新版本chrome怎么安装第三方插件？网上给的方法都用不了，开发模式解压crx，终端安装？ - 知乎用户的回答 - 知乎
https://www.zhihu.com/question/35475402/answer/130750176

**Comment by alexcdever** [comment]
> @alexcdever Mac版最新版本chrome怎么安装第三方插件？网上给的方法都用不了，开发模式解压crx，终端安装？ - 知乎用户的回答 - 知乎
> https://www.zhihu.com/question/35475402/answer/130750176

看来只要是解压了的话，后面的步骤都一样，我是看了一篇专栏介绍的

**Comment by ted423** [comment]
@ted423 我好像记得改Info.plist可以直接装crx,不过我不知道记得有没错

**Comment by ted423** [comment]
@alexcdever https://support.google.com/chrome/a/answer/7517624?hl=en
https://stackoverflow.com/questions/38206916/mandatory-chrome-policies-on-mac
虽然不知道有没用，毕竟我没试过..

**Comment by ccxuy** [comment]
能考虑下支持tempermonkey吗,这样就不用在谷歌商店安装了

**Comment by ted423** [comment]
@ccxuy 提建议是好事，但最好比较清楚的了解你所提建议的利与弊，最起码，tampermonkey上能做到什么程度，你得知道才行。

**Comment by xkww3n** [comment]
> 能考虑下支持tempermonkey吗,这样就不用在谷歌商店安装了

tampermonkey只是一个userscript管理程序，比起chrome扩展来功能受限太多了

**Comment by tainewoo** [comment]
自从微软出了新内核Edge,就不用Chrome好久了.
大家可以考虑更换到Edge.

不过, Edge扩展商店等1.5.1好久了,好像还没过审核?

**Comment by ted423** [comment]
@lbq1121 之后请在这里讨论，issue区一般只提供相关issue的讨论

然后我说下我的看法，我之前说"It's open source",其实我感觉你并没有理解我的意思。
其实开源跟pt也有些类似，你可以把一个开发者理解成一个seeder。
但是seeder并不能保证自己永远在做种，对于leecher来说，可以变成seeder，也可以hit&run。（当然对于开源来说让所有人都变成seeder不太合适），但是leecher面向seeder来说是一个请求，而不是要求。

开源开发者并不保证这个项目的售后，也不保证这个项目的可用性。
因此，issue很可能confirm但没人修，这是正常的。
有人维护，是因为他开心，而不是他有责任和义务去修。

回到release上来,目前的下架很可能跟版权有关，故保持在商店上持续更新，哪怕edge上还活着，意义不大。现有的状态可能是微软没注意到，并不能保证它一直活着。

另一方面open source意味着你可以自己编译，相对于别的语言来说，根据wiki编译还是相当简单的，在没action之前，我都是自己编着用的。
更何况有action之后，直接都是编译好的，除了要能访问github以外，没有任何门槛。wiki也写得相当完善
https://github.com/ronggang/PT-Plugin-Plus/wiki/install-from-crx

并且这个项目不像qb，更新只会减少bug

**Comment by Rhilip** [comment]
写next的时候观察了下mv3以及相关google的政策，
我个人更倾向于我们代码中大量使用eval等可以执行任意代码的字段触发了相关的审核机制。
这在目前的版本中是无法得到优化解决的。另外这影响了我们以下功能：

- 为站点添加自定义插件
![image](https://user-images.githubusercontent.com/13842140/160236957-0f8b10d8-0948-463c-84dd-2601ff953950.png)
![image](https://user-images.githubusercontent.com/13842140/160236962-e505bd87-5b5d-4e31-9c27-a8090b39b058.png)

其他关于站点、下载服务器的在next分支中通过webpack预编译的形式已经能得到一定程度上的解决。

**Comment by emooooooji** [comment]
关注一下进度，能直接在应用商店里安装还是很方便的。

**Comment by ted423** [comment]
https://www.oschina.net/news/196585/fairemail-google

**Comment by PencilCore** [comment]
这下be evil了

**Comment by yuguansu** [comment]
所以google准备自己写各类插件？夺人之鱼不如夺人之渔具？

**Comment by ccxuy** [comment]
我对网页技术不太了解哈，不过eval这个事情貌似能解决？
https://github.com/GrinZero/chrome-inject-eval

**Reply by ted423** [reply]
其实chrome并没有给出具体的解释，都是猜测，个人怀疑其实跟版权有些关系
